protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,131 @@
+/*
+ Copyright 2024 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+
+	"github.com/nats-io/nats.go"
+)
+
+// Protocol is a reference implementation for using the CloudEvents binding
+// integration. Protocol acts as both a NATS client and a NATS handler.
+type Protocol struct {
+
+	// nats connection
+	conn      *nats.Conn
+	url       string
+	natsOpts  []nats.Option
+	connOwned bool
+
+	receiver *receiver
+
+	sender *sender
+}
+
+// NewProtocol creates a new NATS protocol.
+// Use WithURL() OR WithConnection()
+// Use WithConsumerConfig() OR WithOrderedConsumerConfig()
+func NewProtocol(ctx context.Context, sendSubject string, opts ...ProtocolOption) (*Protocol, error) {
+	p := &Protocol{}
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if err := p.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	var errConnection error
+	defer func() {
+		// close connection if an error occurs and we created the nats connection
+		if p.connOwned && p.conn != nil && errConnection != nil {
```

Comment:
Do we need `errConnection != nil` check? Can we not just close the connection if p.conn != nil?

Reply:
Also, do we need to check for p.connOwned if p.conn != nil? Can we not just close the connection then (not sure what p.connOwned actually does)

Reply:
Now I see why I'm confused. See my comments on the Close handling below where this whole logic can be removed from the constructor.